src/util/taproot.rs

```diff
@@ -718,6 +716,37 @@ impl TaprootMerkleBranch {
     }
 }
 
+impl TryFrom<&[sha256::Hash]> for TaprootMerkleBranch {
+    type Error = TaprootError;
+
+    /// Creates a merkle proof from list of hashes.
+    ///
+    /// # Errors
+    ///
+    /// If inner proof length is more than [`TAPROOT_CONTROL_MAX_NODE_COUNT`] (128).
+    fn try_from(v: &[sha256::Hash]) -> Result<Self, Self::Error> {
+        let v = v.to_vec();
```

Comment:
This wasn't my point, the caller can already do this. :) My point was to check the length here so that allocation is avoided in the error case. If you want to avoid duplicating checking logic I have a trick for that:

```rust
// non-public
fn from_collection<T: AsRef<[sha256::Hash]> + Into<Vec<sha256::Hash>>>(collection: T) -> Result<Self, TaprootError> {
    if collection.as_ref().len() > TAPROOT_CONTROL_MAX_NODE_COUNT {
        Err(TaprootError::InvalidMerkleTreeDepth(collection.as_ref().len()))
    } else {
        Ok(TaprootMerkleBranch(collection.into()))
    }
}

// just directly call `Self::from_collection(v)` in all `TryFrom` impls
```

You could also use macros for `TryFrom` and also add `Box<[sha256::Hash]>` as well.

Reply:
Nice trick :) Implemented as suggested, thanks.